pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2084,18 +2084,20 @@ public JSONObject rebalanceTable(final String rawTableName, TableType tableType,
     try {
       RebalanceSegmentStrategy rebalanceSegmentsStrategy =
           RebalanceSegmentStrategyFactory.getInstance().getRebalanceSegmentsStrategy(tableConfig);
-      PartitionAssignment newPartitionAssignment =
-          rebalanceSegmentsStrategy.rebalancePartitionAssignment(idealState, tableConfig, rebalanceUserConfig);
-      IdealState targetState =
-          rebalanceSegmentsStrategy.getRebalancedIdealState(idealState, tableConfig, rebalanceUserConfig);
-
-      // update it as needed
+      // rebalance it as needed
       if (!rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN)) {
         // call the updater providing it a reference to the strategy so it can recompute target state if needed
-        _idealStateUpdater.update(targetState, tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+        _tableRebalancer.rebalance(tableConfig, rebalanceSegmentsStrategy, rebalanceUserConfig);
+        jsonObject.put("status", "Rebalance triggered. Check logs for progress");
+      } else {
+        // dry run mode: return the target partition assignment and idealstate mapping
+        PartitionAssignment newPartitionAssignment =
+            rebalanceSegmentsStrategy.rebalancePartitionAssignment(idealState, tableConfig, rebalanceUserConfig);
+        IdealState targetState =
+            rebalanceSegmentsStrategy.getRebalancedIdealState(idealState, tableConfig, rebalanceUserConfig);
+        jsonObject.put("partitionAssignment", newPartitionAssignment.getPartitionToInstances());
+        jsonObject.put("idealState", targetState.getRecord().getMapFields());
```

Comment:
not part of this PR, but this is method is mixing the responsibility of controllerRestResource and PinotHelixResourceManager. Returning a json object from this method is not a good idea. Also, this should be sync and we can make it async from the RestApi if needed.